src/base/serial/fossil.c

```diff
@@ -351,30 +352,59 @@ void fossil_int14(int num)
   }
 
   /* Get FOSSIL driver info. */
-  case 0x1b:
-  {
+  case 0x1b: {
     unsigned char *p = SEG_ADR((unsigned char *), es, di);
-    int ifree = FOSSIL_RX_BUFFER_SIZE - RX_BUF_BYTES(num),
-      ofree = FOSSIL_TX_BUFFER_SIZE - serial_get_tx_queued(num),
-      bufsize = (LWORD(ecx) <= 19) ? LWORD(ecx) : 19;
-    if (ofree < 0)
-      ofree = 0;
-    if (LO(dx) == 0xff)
-    {
-      fossil_check_info();
+    int ifree, ofree, bufsize;
+    fossil_info_t fossil_info, *fi;
+
+    if (!fossil_tsr_installed) {
+      LWORD(eax) = 0;       // Perhaps the only way to indicate no valid data
       return;
     }
-    /* Fill in some values that aren't constant. */
-    com[num].fossil_info[10] = ifree & 0xff;
-    com[num].fossil_info[11] = ifree >> 8;
-    com[num].fossil_info[14] = ofree & 0xff;
-    com[num].fossil_info[15] = ofree >> 8;
+
+    bufsize = (LWORD(ecx) < sizeof(*fi)) ? LWORD(ecx) : sizeof(*fi);
+
+    if (LO(dx) == 0xff) {
+      fi = &fossil_info;
+
+      fi->size = sizeof(*fi);
+      fi->frev = FOSSIL_REVISION;
+      fi->irev = 0;         // Driver revision (not used)
+      fi->id_offset = fossil_id_offset;
+      fi->id_segment = fossil_id_segment;
+      fi->rx_bufsize = FOSSIL_RX_BUFFER_SIZE;
+      fi->rx_remaining = 0; // Not valid in non port specific context
+      fi->tx_bufsize = FOSSIL_TX_BUFFER_SIZE;
+      fi->tx_remaining = 0; // Not valid in non port specific context
+      fi->scrn_width = 80;
+      fi->scrn_height = 25;
+      fi->bps = 0;          // Bps rate (not used)
+
+#if SER_DEBUG_FOSSIL_STATUS
+      s_printf("SER: FOSSIL 0x1b: Driver info, ibuf=%d, obuf=%d, "
+               "idaddr=%04x:%04x, AX=%d\n",
+               FOSSIL_RX_BUFFER_SIZE, FOSSIL_TX_BUFFER_SIZE, fossil_id_segment,
+               fossil_id_offset, bufsize);
+#endif
+    } else {
+      fi = &com[num].fossil_info;
+
+      /* Fill in some values that aren't constant. */
+      ifree = FOSSIL_RX_BUFFER_SIZE - RX_BUF_BYTES(num);
+      ofree = FOSSIL_TX_BUFFER_SIZE - serial_get_tx_queued(num);
```

Comment:
No because it is not a linux-exported API
that we use. We shouldn't use every random
define from linux headers; only those that
we later feed to the linux-provided API.
Otherwise you make serial port emulation
depending on linux and unportable to, say,
BSD. There is a tty_io.c that communicates
to linux, everything else should not include
any linux header as being platform-independent.